### PR TITLE
Use basestore in repos store

### LINKED ISSUE #14754: repo-store: Convert the repos store to use basestore
Convert the [internal/db/repos](https://github.com/sourcegraph/sourcegraph/blob/938cca4cb22b2fbaf9399c7164566a8337bb935a/internal/db/repos.go#L42-L42) store to extend the [basestore.Store](https://github.com/sourcegraph/sourcegraph/blob/645c7cb65eb6f639d5eaf5b6209015413f1d0ce2/internal/db/basestore/store.go#L38-L38)

Extracted from https://github.com/sourcegraph/sourcegraph/issues/14092 and https://github.com/sourcegraph/sourcegraph/issues/11676

### PR DESCRIPTION
**Pull request by asdine**
It also copies all the `repos.Repo` and `repos.Repos` helpers to the `cmd/frontend/types` package for further use.
The original methods will be cleaned up in future PRs.

Fixes #14754

### CODE REVIEW

File: internal/db/repos.go [hunk 1]
```diff
@@ -47,18 +49,49 @@ func (e *RepoNotFoundErr) NotFound() bool {
 	return true
 }
 
-// repos is a DB-backed implementation of the Repos
-type repos struct{}
+// ReposStore is a DB-backed implementation of the Repos.
+type ReposStore struct {
```
Comment (ryanslade) [inline internal/db/repos.go:53]: nit: I think `RepoStore` sounds better and avoid the double s.

File: internal/db/repos.go [hunk 2]
```diff
@@ -47,18 +49,49 @@ func (e *RepoNotFoundErr) NotFound() bool {
...
+	*basestore.Store
+
+	m sync.Mutex
```
Comment (ryanslade) [inline internal/db/repos.go:56]: nit: We usually use `mu` for mutexes

File: internal/db/repos.go [hunk 3]
```diff
@@ -47,18 +49,49 @@ func (e *RepoNotFoundErr) NotFound() bool {
...
 // only from the database and NOT from any external sources. If the
 // caller is concerned the copy of the data in the database might be
 // stale, the caller is responsible for fetching data from any
 // external services.
-func (s *repos) Get(ctx context.Context, id api.RepoID) (*types.Repo, error) {
+func (s *ReposStore) Get(ctx context.Context, id api.RepoID) (*types.Repo, error) {
 	if Mocks.Repos.Get != nil {
 		return Mocks.Repos.Get(ctx, id)
 	}
+	s.ensureStore()
```
Comment (ryanslade) [inline internal/db/repos.go:94]: Maybe we should create a follow up issue to stop using `dbconn.Global`?

File: internal/db/repos.go [hunk 4]
```diff
@@ -164,7 +201,8 @@ func (s *repos) Count(ctx context.Context, opt ReposListOptions) (ct int, err er
 	tr.LazyPrintf("SQL: %v", q.Query(sqlf.PostgresBindVar))
 
 	var count int
-	if err := dbconn.Global.QueryRowContext(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...).Scan(&count); err != nil {
+
+	if err := s.Store.Handle().DB().QueryRowContext(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...).Scan(&count); err != nil {
```
Comment (ryanslade) [inline internal/db/repos.go:205]: Maybe worth exposing `QueryRow` directly on basestore?

**Review by ryanslade — APPROVED** [review]
Nice 👍 

Looking forward to the end result here after the cleanups

File: internal/db/repos.go (hunk 3 above)
Comment (asdine) [reply internal/db/repos.go:94]: Good idea, here it is: https://github.com/sourcegraph/sourcegraph/issues/15092

**Pull request merged**